### ISSUE TITLE
Is it possible to explicitly name a dynamic?

### ISSUE THREAD

**Issue by russellseymour** [sparkleformation/sparkle_formation#177]
I am using SparkleFormation to build up an Azure template and I am pleased with the results.

However one of the things that needs to be specified in the `storageaccount`.  There is a built in dynamic for this but it follows the normal convention of appending the type to the name that it is given, e.g.

```
dynamic!(:storage_accounts, "456sfjgk6")
```

Will give the resource a name of `456sfjgk6storageAccounts`.

Microsoft state that the storage account must be globally unique (as it forms part of the DNS name) and must be between 3 and 24 characters.  If the last 16 characters are always the same then this will reduce the entropy available to make it unique.

Most templates allow a variable to be set that sets the name of the storageAccount, however it is not possible to reference the variable as the name.  This would be useful for _all_ resources as it makes the template reusable without Sparkle.

So is there a workaround I can use to achieve what I want, which is to end up with something like the following as the resultant JSON?

```
    {
      "type": "Microsoft.Storage/storageAccounts",
      "name": "[variables('storageAccountName')]",
      "apiVersion": "2015-06-15",
      "location": "[resourceGroup().location]",
      "properties": {
        "accountType": "[parameters('storageAccountType')]"
      }
    }
```

Additionally storage account must only consist of numbers and lower case letters only.

Please let me know if you need more information as this would really help with the composing of ARM templates.

Thank you, Russell

**Comment by jjasghar** [comment]
@chrisroberts @calebhailey @cwjohnston any chance ya'll could take a look at this? 

@russellseymour and I are working in Azure more and more this would help unblock us on somethings.

**Comment by luckymike** [comment]
Hey guys, 

You can do this with the `:resource_name_suffix` key, like so:
```
dynamic!(:storage_accounts, "456sfjgk6", :resource_name_suffix => :acct)
```
Which should give you the name `456sfjgk6Acct`, which is shorter. I'm not sure if you can use an empty string for the suffix, but you could certainly go down to just 1 or 2 letters.

This is in the [dynamic documentation]9http://www.sparkleformation.io/docs/sparkle_formation/building-blocks.html#dynamics) right above "Dynamic Return Context"

**Comment by russellseymour** [comment]
@luckymike Thanks for this, that is a great help.  Now I understand how the `:resource_name_suffix` can be used :-).

I am trying to make it so that the ARM template that SparkleFormation generates can be reused as a normal Azure template.  To do that the name for resources needs to be specified as a reference to a parameter or a variable, e.g.:

```
"name": "[variables('storageAccountName')]",
```

Is there anyway that this can be done using Sparkle?  I am guessing that out of the box this is not supported because of the internal referencing.

**Comment by chrisroberts** [comment]
Hi! Something that I added for GCE specifically was a custom option for the `dynamic!` method called `:sparkle_unique` that will generate repeatably unique resource names due to the lack of logical resource id vs physical resource id on things like buckets:

https://github.com/sparkleformation/sparkle_formation/blob/master/lib/sparkle_formation/sparkle_attribute/google.rb#L59

As this is apparently an issue on other providers as well, I'll be moving that into the core functionality so it can be utilized under any provider.

As for referencing variables as above, sure, you can do that easily:

```ruby
name variables!(:storage_account_name)
```

**Comment by russellseymour** [comment]
@chrisroberts Something like `:sparkle_unique` will be useful when a template is being generated for use there and then.  However I am trying to use Sparkle to build up reusable ARM templates.  An example will make this easier to explain.

The following ARM template simply creates a `storageAccount` in Azure.

```json
{
  "$schema": "http://schema.management.azure.com/schemas/2014-04-01-preview/VM_IP.json",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "storageAccountName": {
      "type": "string"
    },
    "storageAccountType": {
      "type": "string",
      "defaultValue": "LRS"
    }
  },
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "name": "[parameters('storageAccountName')]",
      "apiVersion": "2015-06-15",
      "location": "[parameters('location')]",
      "properties": {
        "accountType": "[parameters('storageAccountType')]"
      }
    }
  ]
}
```

As can be seen the name of the resource is based on the parameter that has been specified.  In Sparkle's current form if I set the name of the resource to be:

```
dynamic!(:storageAccount, "[parameters('storageAccountName')]")
```

The output for the name of the resource is `[parameters('storageAccountName')]StorageAccounts` which is not valid.  I have tried @luckymike suggestion of using the `:resource_name_suffix` but strangely it does not change the name, e.g.:

```
dynamic!(:storageAccount, "[parameters('storageAccountName')]", :resource_name_suffix => :acct)
```

Results in `[parameters('storageAccountName')]StorageAccounts`, which seems to suggest the `:resource_name_suffix` is not being honoured.

I need a way to be able to explicitly set the name of a resource in a `dynamic`.  One that is not appended to and the system will just use that string, e.g.

```
dynamic!(:storageAccount, "[parameters('storageAccountName')]", :literal => true)
```

When something is then referenced, e.g. in a `depends_on` then it would just be the string and the `.to_sym` would not be required:

```
depends_on!("[concat('Microsoft.Storage/storageAccounts/', parameters('storageAccountName')]")
```

Which correctly then generates the following JSON snippet

```
"dependsOn": [
    "[concat('Microsoft.Storage/storageAccounts/', parameters('storageAccountName')]"
]
```

Can or could this be done?

Thanks.

**Comment by chrisroberts** [comment]
Thanks for the expanded use case. Yes, this can be done. Also, defining intrinsic functionality within strings is not required within sparkleformation. You can write it as Ruby using helper methods and it will produce the value in the expected string format. A simple version of the above use case:

```ruby
SparkleFormation.new(:azure, :provider => :azure) do

  dynamic_bucket = dynamic!(
    :storage_accounts,
    parameters!(:some_name).to_s,
    :resource_name_suffix => nil
  )

  dynamic!(:storage_accounts, :concat_depends_on) do
    depends_on! parameters!(:some_name)
  end

  dynamic!(:storage_accounts, :reference_name) do
    depends_on! dynamic_bucket.resource_name!
  end

end
```
which results in:
```json
{
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2016-01-01",
      "location": "[resourceGroup().location]",
      "name": "[parameters('someName')]"
    },
    {
      "type": "Microsoft.Storage/storageAccounts",
      "dependsOn": [
        "[parameters('someName')]"
      ],
      "apiVersion": "2016-01-01",
      "location": "[resourceGroup().location]",
      "name": "concatDependsOnStorageAccounts"
    },
    {
      "type": "Microsoft.Storage/storageAccounts",
      "dependsOn": [
        "[parameters('someName')]"
      ],
      "apiVersion": "2016-01-01",
      "location": "[resourceGroup().location]",
      "name": "referenceNameStorageAccounts"
    }
  ]
}
```
The important bits being the `.to_s` on the helper call and the `nil` value for the suffix. Having a `:literal` option to automatically do the `to_s` + `nil` would be nice, so I'll add that in on the next release.

The intrinsic functions available within the templates are available within sparkleformation so you can build complex things without having to write it as a string of code. For example, your `concat` example above can be written like:

```ruby
depends_on! concat!('Microsoft.Storage/storageAccounts/', parameters!(:storage_account_name))
```
which results in:
```json
{
  "dependsOn": [
    "[concat('Microsoft.Storage/storageAccounts/', parameters('storageAccountName'))]"
  ]
}
```
Cheers!

**Comment by russellseymour** [comment]
@chrisroberts Thanks very much for this - it works perfectly :-)

I have another question, but it is on a slight different topic so I will open another issue.
